File: backend/app/schemas/admin_overview.py
from datetime import datetime

from pydantic import BaseModel


class StatCardValue(BaseModel):
    current: int
    previous: int | None
    delta_pct: float | None
    sparkline: list[int]


class ActivityEntry(BaseModel):
    description: str
    occurred_at: datetime


class RequestPreview(BaseModel):
    id: str
    dataset_title: str
    requester_name: str
    submitted_at: datetime


class StorageUsageSchema(BaseModel):
    used_bytes: int
    capacity_bytes: int | None
    percent_used: float | None


class SystemHealthSchema(BaseModel):
    database: bool
    redis: bool
    storage: bool
    checked_at: datetime


class AdminOverviewResponse(BaseModel):
    total_users: StatCardValue
    active_users: StatCardValue
    total_datasets: StatCardValue
    downloads: StatCardValue
    pending_requests: StatCardValue
    recent_requests: list[RequestPreview]
    recent_activity: list[ActivityEntry]
    category_distribution: dict[str, int]
    storage: StorageUsageSchema
    top_downloaded: list[dict]
    system_health: SystemHealthSchema


class StorageCapacitySchema(BaseModel):
    storage_capacity_bytes: int | None

    model_config = {"from_attributes": True}


class StorageCapacityUpdate(BaseModel):
    storage_capacity_bytes: int | None = None

File: frontend/src/lib/types/admin-overview.ts
// Mirrors backend/app/schemas/admin_overview.py exactly — keep these two in sync.

export interface StatCardValue {
  current: number;
  previous: number | null;
  delta_pct: number | null;
  sparkline: number[];
}

export interface ActivityEntry {
  description: string;
  occurred_at: string;
}

export interface RequestPreview {
  id: string;
  dataset_title: string;
  requester_name: string;
  submitted_at: string;
}

export interface StorageUsageSchema {
  used_bytes: number;
  capacity_bytes: number | null;
  percent_used: number | null;
}

export interface SystemHealthSchema {
  database: boolean;
  redis: boolean;
  storage: boolean;
  checked_at: string;
}

export interface TopDownloadedEntry {
  dataset_title: string;
  count: number;
}

export interface AdminOverviewResponse {
  total_users: StatCardValue;
  active_users: StatCardValue;
  total_datasets: StatCardValue;
  downloads: StatCardValue;
  pending_requests: StatCardValue;
  recent_requests: RequestPreview[];
  recent_activity: ActivityEntry[];
  category_distribution: Record<string, number>;
  storage: StorageUsageSchema;
  top_downloaded: TopDownloadedEntry[];
  system_health: SystemHealthSchema;
}

export interface StorageCapacitySchema {
  storage_capacity_bytes: number | null;
}

export interface StorageCapacityUpdate {
  storage_capacity_bytes?: number | null;
}
